.travis.yml

```diff
@@ -13,74 +18,118 @@ python:
 
 services:
   - postgresql
-  - neo4j
-  - docker
 
-install:
-  - node --version
-  - NODE_VERSION=6.11.0
-  - nvm install $NODE_VERSION
-  - nvm use $NODE_VERSION
-  - node --version
-
-  - npm --version
-  - npm install -g npm@3.10.10
-  - npm --version
-
-  - pip install -r requirements.txt --quiet
-  # Install geckodriver required for selenium testing
-  - GECKODRIVER_VERSION=v0.17.0
-  - TAR=geckodriver-$GECKODRIVER_VERSION-linux64.tar
-  - wget https://github.com/mozilla/geckodriver/releases/download/$GECKODRIVER_VERSION/$TAR.gz
-  - gunzip $TAR.gz
-  - tar -xvf $TAR
-  - chmod a+x geckodriver
-  - export PATH=$PATH:`pwd`
 env:
   global:
     # These env vars are available to every build
     - PYTHONPATH=$PYTHONPATH:../refinery:../refinery/config
+    - CODECOV=codecov
     - DJANGO_SETTINGS_MODULE=config.settings.prod
+    - FLAKE8=flake8
+    - GRUNT=grunt
+    - SELENIUM_TESTS=selenium-tests
+
+matrix:
```

Comment:
Why not just use a single variable for running different test suites as described here: https://docs.travis-ci.com/user/speeding-up-the-build/
This way there is no need for all the conditionals.

Reply:
@hackdna Wouldn't we still need conditionals even if we used a single variable w/ different values? I'm not seeing how we could benefit from that here

Reply:
This is the whole point of the build matrix.

Reply:
If you could elaborate and alert me as to how we could utilize the build matrix and not need conditional statements in our `travis.yaml` that would be very helpful

Otherwise I'm still seeing a need for conditional statements

Reply:
Some more examples:
https://docs.travis-ci.com/user/customizing-the-build/#Build-Matrix
https://docs.travis-ci.com/user/build-stages/matrix-expansion/

Something like:
```yaml
env:
  - TEST_SUITE="grunt make"
  - TEST_SUITE="manage.py test"
script: $TEST_SUITE
```
"[Travis CI will determine the build matrix based on the environment variables and schedule two builds to run](https://docs.travis-ci.com/user/speeding-up-the-build/#Parallelizing-your-builds-across-virtual-machines)"

Reply:
Yes, but for example Selenium tests are dependent on grunt operations while Django tests are not. 

So I don't believe this approach fits our use case unless I am mistaken.

Reply:
Does it matter?

Reply:
I mean, the whole idea behind the conditionals is that we don't run unnecessary code for builds that don't need it. No downloading of `geckodriver` if we don't need to, no compilation of static assets if we don't need to etc.

So I would say that it does matter. Otherwise splitting up the builds doesn't make much sense if we are going to run the majority of the same code in each and just have a different test command to execute.

I'm not convinced that the approach you're suggesting gains us anything, but I'm looking to be swayed.

Reply:
TBH, I'm not even sure if parallelizing the build gains us much while it certainly makes `.travis.yml` a lot more complicated. How much time is actually saved compared to a regular build? If you look at the [build history](https://travis-ci.org/refinery-platform/refinery-platform/builds) the gains are not clear at all.

Reply:
So [this parallelized build](https://travis-ci.org/refinery-platform/refinery-platform/builds/286079607) took about 12 minutes

While [this non-parallelized build](https://travis-ci.org/refinery-platform/refinery-platform/builds/286116497) took about 19 minutes

The `total time` that Travis reports back for the parallelized builds seems to just be an addition of all the builds time not the actual wall clock time

Reply:
I see. Still, there is a lot of variation in run times: this [regular build](https://travis-ci.org/refinery-platform/refinery-platform/builds/286137728) took 13 mins 39 secs while this [parallelized build](https://travis-ci.org/refinery-platform/refinery-platform/builds/286048105) ran for 13 mins 37 secs.

Also, the time waiting to run a build is not taken into account and that can vary a lot, so benefits are certainly not clear, especially given the additional code complexity. If a quick test run is required, it is very fast to do that locally and can be done at different levels of granularity.

I don't think installing dependencies for each build job would impact the runtime significantly since a lot of them are now cached but it would make `.travis.ci` a lot simpler.

To make a long story short, I think it's worth to parallelize the builds only if it doesn't lead to a significant increase in complexity of the build code (i.e., use a standard build matrix expansion and avoid conditionals).

Reply:
The parallelized build you referenced was before I split the selenium builds into their own respective builds so that won't be an accurate time estimate.

I don't think the time a build is pending before its run is relevant in this conversation.

I still see benefit in isolating code specific to each build regardless of how its done. For example, if there is a failed migration, and we run our migrations in every parallelized build we see 4 failed builds and then have to dive deeper to figure out what went wrong. I was going to add shell scripts to make `.travis.yml` a bit more tidy looking but even then we have ~5 more things to maintain.

As for code complexity, we are adding some if statements to get general gains of ~40% speed per build. I find that to be an okay tradeoff.

Reply:
The point is that there is large inherent variation in run times that we can't control. Also, the wait time is very relevant because we are interested in the total run time: it often takes the same time (or longer) to wait for a build to start running as it does to run. All this makes the build time savings of a few minutes marginal (see also: http://wiki.c2.com/?ProfileBeforeOptimizing).

Moreover, the build times on Travis should not even matter. It is not a tool for rapid iteration - your local environment is.

Btw, if the build is not parallelized then any error like a failed migration would result in a single failed build :)

Finally, if you really want to make Travis code more robust, it would be more useful to [make it use Puppet manifests](https://github.com/refinery-platform/refinery-platform/issues/822). No need for custom scripts that duplicate existing functionality.